system/globalid/lnuid.cpp

```diff
@@ -0,0 +1,403 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef __linux__
+#include <sys/time.h>
+#elif _WIN32
+#include <windows.h>
+#include <wincrypt.h>
+#include <chrono>
+#else
+
+#endif
+
+/**
+* This is an implementation of Globally Unique Transaction ID’s.
+* Note: This class currently generates unique id on OS's that supports device /dev/urandom.
+* @author - Amol Patwardhan
+* Version: 1.0
+*/
+
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include <vector>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <sstream>
+#include <ctime>
+#include "lnuid.h"
+using namespace std;
+
+namespace ln_uid {
+
+    /** All alphanumeric characters except for "0", "I", "O", and "l" */
+    const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+    //Random byte count
+    const unsigned int random_byte_count = 11;
+    //Random byte count
+    const unsigned int time_byte_count = 5;
+
+    //NOTE: This variable is used to offset dropped 41st bit from time component.
+    //      This variable adds time till last 41st bit flip on 03-Nov-2004.
+    //      This value should be updated later to handle next bit flip on 07-Sep-2039.
+    unsigned long long  int OFFSET = (unsigned long long  int) 1099511627776;
+
+
+    ln_uid_t &createUniqueId(ln_uid_t &out){
+
+        unsigned long long  int millisecondsSinceEpochTrim;
+#ifdef __linux__
+        struct timeval tv;
+        gettimeofday(&tv, NULL);
+
+        millisecondsSinceEpochTrim =
+        (unsigned long long int)(tv.tv_sec) * 1000 +
+        (unsigned long long int)(tv.tv_usec) / 1000;
+#elif _WIN32
+        millisecondsSinceEpochTrim =
+        std::chrono::system_clock::now().time_since_epoch() /
+        std::chrono::milliseconds(1);
+#else
+
+#endif
+
+
+
+        unsigned char timedata[time_byte_count];
+
+        timedata[0] = (int)((millisecondsSinceEpochTrim >> 32) & 0xFF);
+        timedata[1] = (int)((millisecondsSinceEpochTrim >> 24) & 0xFF);
+        timedata[2] = (int)((millisecondsSinceEpochTrim >> 16) & 0xFF);
+        timedata[3] = (int)((millisecondsSinceEpochTrim >> 8) & 0XFF);
+        timedata[4] = (int)((millisecondsSinceEpochTrim & 0XFF));
+
+        unsigned char randomdata[random_byte_count];
+
+
+#ifdef __linux__
+        FILE *fp;
+        fp = fopen("/dev/urandom", "r");
+        fread(&randomdata, 1, random_byte_count, fp);
+        fclose(fp);
+#elif _WIN32
+        HCRYPTPROV hProvider;
+        CryptAcquireContext(&hProvider, 0, 0, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT);
+        CryptGenRandom(hProvider, random_byte_count, randomdata);
+#else
+
+#endif
+
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            out[i] = timedata[j];
```

Comment:
Why do you need both i and j (they are the same)?

Reply:
Agreed.  Copied from Amol, but I'll change it.